src/main/network/dns.rs

```diff
@@ -0,0 +1,170 @@
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Write;
+use std::net::Ipv4Addr;
+use std::os::fd::AsRawFd;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use rustix::fs::MemfdFlags;
+use shadow_shim_helper_rs::HostId;
+
+#[derive(Debug)]
+struct Database {
+    name_index: HashMap<String, Arc<Record>>,
+    addr_index: HashMap<Ipv4Addr, Arc<Record>>,
+}
+
+#[derive(Debug)]
+struct Record {
+    id: HostId,
+    addr: Ipv4Addr,
+    name: String,
+}
+
+#[derive(Debug)]
+pub struct DnsBuilder {
+    db: Database,
+}
+
+impl DnsBuilder {
+    pub fn new() -> Self {
+        Self {
+            db: Database {
+                name_index: HashMap::new(),
+                addr_index: HashMap::new(),
+            },
+        }
+    }
+
+    pub fn register(&mut self, id: HostId, addr: Ipv4Addr, name: String) {
+        if !addr.is_loopback() && !addr.is_unspecified() {
+            let record = Arc::new(Record {
+                id,
+                addr,
+                name: name.clone(),
+            });
+            self.db.name_index.insert(name, record.clone());
+            self.db.addr_index.insert(addr, record);
+        }
+    }
+
+    pub fn into_dns(self) -> anyhow::Result<Dns> {
+        let pid = std::process::id();
+        let name = format!("shadow_dns_hosts_file_{pid}");
+
+        let memfd = rustix::fs::memfd_create(name, MemfdFlags::CLOEXEC)?;
+        let mut file = File::from(memfd);
+
+        // Sort the records to produce deterministic ordering in the hosts file.
+        let mut records: Vec<&Arc<Record>> = self.db.addr_index.values().collect();
+        records.sort_by(|a, b| a.addr.cmp(&b.addr));
+
+        writeln!(file, "127.0.0.1 localhost")?;
+        for record in records.iter() {
+            writeln!(file, "{} {}", record.addr, record.name)?;
+        }
+
+        let path = PathBuf::from(format!("/proc/{pid}/fd/{}", file.as_raw_fd()));
```

Comment:
Well, I'm not sure that the workers that would end up using the path have the same PID as the process that created the `memfd`. If they don't have the same PID, would the FD still be accessible?

(Also, I followed the example that Jim implemented in process, but I'm happy to change if we think we can get away without storing the PID.)

Reply:
I think it would work since the workers and main process all have the same pid but different tid's. There's `/proc/self` which expands to `/proc/<pid>`, which is separate from `/proc/thread-self` which expands to `/proc/self/task/<tid>`.

I don't think processes can access `/proc/<pid>/fd/<fd>` from other processes anyways (except maybe following the ptrace rules where parent processes can manage child processes).